### PR TITLE
Drop piped table env_ids index

### LINKED ISSUE #3191: Adding second piped on the control plain hides piped from list
**What happened**:
Adding first piped works and show up in the list.
But adding second piped hides first and second piped from list.

Note: This happends in case of using MySQL as datastore in quickstart environment

**What you expected to happen**:
First piped and second piped are displayed in the screen when second piped is added.

**How to reproduce it**:
```
helm repo add pipecd https://charts.pipecd.dev

helm install pipecd pipecd/pipecd -n pipecd \
  --create-namespace \
  --dependency-update \
  --values https://raw.githubusercontent.com/pipe-cd/manifests/v0.25.0/quickstart/control-plane-values.yaml

# Add 2 piped from control plain
```

**Environment**:
- `piped` version: none
- `control-plane` version: v0.25.0
- Others:

### PR DESCRIPTION
**Pull request by khanhtc1202**
**What this PR does / why we need it**:

This PR removes the unneeded index `piped_project_id_env_ids_asc` on Piped table. The queries on MySQL datastore work with/without index but we chose to make that index since this `EnvIds` field is a virtual generated column, which its value will be generated every time we query for it, which makes the operation a bit slower.
Since the `Environment` concept is deprecated and will be removed in the near future, I think it's fine to just remove this index to get rid of environment-related unstable queries issues. Besides, the number of piped(s) which be controlled by a control-plane is not too many, query without index is acceptable at this meantime.

Tested on my local, some places we needed to use this index previously

Screen Shot 2022-02-06 at 14 20 25

Screen Shot 2022-02-06 at 14 21 17

**Which issue(s) this PR fixes**:

Fixes #3191

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by knanao** [comment]
Nice catch!
I thought it's not too late to remove `Environment` completely, but what do you think about it?

**Comment by khanhtc1202** [comment]
> I thought it's not too late to remove `Environment` completely, but what do you think about it?

Sorry, not really get your point, could you please explain why does it's too late to remove environment concept completely 👀

**Comment by knanao** [comment]
Sorry for lack of explanation.
I didn't understand why to remove it even though it's still used.
That's why, I thought it's better that drop it after remove `Environment` completely.

**Comment by khanhtc1202** [comment]
> I didn't understand why to remove it even though it's still used.

I see, the problem is that MySQL can work without this index, but as I mentioned in the PR description, we chose to make this index because it's a trick to make the queries on virtual generated column (like this EnvIds column) faster. Without this index, the queries we use still work but a bit slower (depending on the size of the table). At this point of time, due to issue #3191, I realized that instead of workaround with default value for this column to avoid problems (as we did in PR https://github.com/pipe-cd/pipecd/pull/3164), I think it's easier to just remove this index (the statement which used to generate this EnvIds column is remained, just remove the index), and we will remove this deprecated Environment anyway, so this fix is good enough.

**Comment by knanao** [comment]
> Without fixing this (wait until we remove the Environment concept completely), MySQL datastore users will not about to create more than one piped without environment value, which is much annoying to them. So I think, a quick and effective fix is needed.
Oh, That's make sense.
Thank you for your explanations.

/lgtm

File: pkg/datastore/mysql/ensurer/indexes.sql [hunk 1]
```diff
@@ -124,6 +124,10 @@ CREATE INDEX piped_project_id_asc ON Piped (ProjectId);
 -- index on `ProjectId` ASC and `EnvIds` ASC
 ALTER TABLE Piped ADD COLUMN EnvIds JSON GENERATED ALWAYS AS (IFNULL(data ->> "$.env_ids", '[]')) VIRTUAL NOT NULL;
 CREATE INDEX piped_project_id_env_ids_asc ON Piped (ProjectId, (CAST(EnvIds AS CHAR(36) ARRAY)));
+-- Remove the `piped_project_id_env_ids_asc` index due to issue around empty array value on column
+-- which be a part of mulitple columns index.
+-- ref: https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-multi-valued
+DROP INDEX piped_project_id_env_ids_asc ON Piped;
```
Comment (nghialv) [inline pkg/datastore/mysql/ensurer/indexes.sql:130]: Does this return an error when the index has already been removed before?
Do we have to check its existence before dropping it?

Comment (khanhtc1202) [reply pkg/datastore/mysql/ensurer/indexes.sql:130]: Nice question! In fact, there is no straightforward way to check like `DROP INDEX IF EXISTS` in MySQL 8 😢 There's some workaround available for that, but I chose to keep the `CREATE INDEX` statement as is and add this `DROP INDEX` statement after that, which means
- if this is the first time this ops runs on a newly created datastore, the index will be created first, then remove
- if this is NOT the first time this ops runs on in a used datastore, the `CREATE INDEX` yield `Duplicated` error which will be ignored (handled already), and then this `DROP INDEX` statement is run to remove that existing index
That's the current situation, plz lets me know if you have a better idea for that 🙏

Comment (nghialv) [reply pkg/datastore/mysql/ensurer/indexes.sql:130]: Thanks for your deep explanation. I got your point.

How about removing the `CREATE INDEX` and ignoring the error from `DROP INDEX`?

Comment (khanhtc1202) [reply pkg/datastore/mysql/ensurer/indexes.sql:130]: That's fine too, then let me update the ensurer to ignore that. TBH, I don't like the current "create then remove" way too, but just think that there're some reasons I don't know that make them abandon the `DROP INDEX IF EXISTS`, and want to return the error on `DROP INDEX` explicitly.

Comment (khanhtc1202) [reply pkg/datastore/mysql/ensurer/indexes.sql:129]: Thanks for the idea, addressed by https://github.com/pipe-cd/pipecd/pull/3206/commits/895850ac050fb34f1a8f1ca472e5019853b7f3f8

**Comment by nghialv** [comment]
Thank you.
/lgtm

**Comment by knanao** [comment]
Cool!
/approve

**Pull request merged**